### PR TITLE
make the router module great again

### PR DESCRIPTION
**Pull request by Guria**
For testing, the router from this branch can be installed with:

```
npm install cerebral/cerebral#great-router
```

TODO:
* [x] run router.routed signal with bound mutations
* [ ] investigate recorder integration (recorder should store and replay not only matations, but signals too)
* [ ] getSignalUrl
* [ ] getPathUrl
* [ ] setActivePathUrl
* [x] "computed" to calculate urls (gaspard)
* [x] tests (gaspard)
* [x] support tags in route
* [x] check history back/forward (gaspard)

### CODE REVIEW

File: packages/cerebral-router/src/router.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    }
+  }
+
+  onFlush(changed) {
+    if (!this.activeStateMap) {
+      return
+    }
+    const {mapByTarget, route} = this.activeStateMap
+    if (
+      Object.keys(mapByTarget.state).find((key) => {
```
Comment (Guria) [inline packages/cerebral-router/src/router.js]: @christianalfoni do we have some utility to determine if some path is changed?

Comment (christianalfoni) [inline packages/cerebral-router/src/router.js]: @Guria No utility, but without gotten a chance to review code I would suggest converting every state dependency path to an object of:

```js
{
  foo: true,
  bar: {
    baz: true
  }
}
```

And when you get the `change` object, which is similar, you traverse it and see if you hit something. 

But yeah, it is only used once inside dependency store which both components and computed uses, so did not create a utility for it

File: packages/cerebral-router/src/router.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+      event && event.preventDefault()
+
+      if (signal) {
+        if (map && prevSignal && prevSignal.name === signal && !propsDiffer(input, prevSignal.input)) {
+          // do not re run signal if input not changed
+          return
+        }
+
+        console.log('start signal from url change')
```
Comment (osv) [inline packages/cerebral-router/src/router.js]: debug?

File: packages/cerebral-router/src/router.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            }
+
+            resolved[key] = map[key](ctx).value
+            return resolved
+          }, {})
+          : payload
+      )
+
+      this.setUrl(url)
+      console.log('update on signal start')
```
Comment (osv) [inline packages/cerebral-router/src/router.js]: again debug?

**Comment by edgesoft** [comment]
#683

**Comment by gaspard** [comment]
I am working on fixing tests in matching.test.js. I added some code for compatibility to use previous route definition format.

File: packages/cerebral-router/README.md [hunk 4]
```diff
@@ -71,6 +74,58 @@ const controller = Controller({
...
+
+When you need more advanced behavior, you define routes with an array of
+route definitions like this:
+
+```js
+import {Controller} from 'cerebral'
+import Router from 'cerebral-router'
+
+const controller = Controller({
+  router: Router({
```
Comment (henri-hulski) [inline packages/cerebral-router/README.md]: `router` should be a module.

Comment (gaspard) [inline packages/cerebral-router/README.md]: Fixed, also changed import statement using new module naming.

**Comment by henri-hulski** [comment]
Can TODO.md be removed?

**Comment by henri-hulski** [comment]
`filterFalsy` option is missing in README.

**Comment by gaspard** [comment]
OK, the router is rebased on top @reflog bug fix. All tests pass and I checked todomvc and demo.

**Comment by Guria** [comment]
Thanks guys. Will review and test it soon.

**Comment by Guria** [comment]
@cerebral/collaborators  Force pushed rebased branch

**Comment by gaspard** [comment]
Something went wrong during the rebase: lots of files are missing:

```
	src/base.js
	src/operators/reload.js
	src/router.baseUrl.test.js
	src/router.js
	src/router.onlyHash.test.js
	src/router.provider.test.js
	src/router.stateMapping.test.js
	src/router.urlMapping.test.js
	src/utils.test.js
```

Any idea what happened ?

If there is no reason for this, I'll just move the files back in and make sure tests pass.

**Comment by henri-hulski** [comment]
I can look into it in the evening. At the moment I'm on work. Have some carpentry at the moment.

**Comment by Guria** [comment]
Hmm. Looks really strange. I can checkout this files from branch before rebase. Is list contains all missing files?

**Comment by gaspard** [comment]
Yes... list is complete.

**Comment by Guria** [comment]
they are all left in cerebral-router folder

**Comment by gaspard** [comment]
Ah OK, I'll just move them then.

**Comment by henri-hulski** [comment]
So I forgot to move them during rebase. I moved some files (from forms and others) but didn't see the cerebral-router folder. I fixed the others in rename commit (https://github.com/cerebral/cerebral/commit/7ba1f2e76802c1eb4f155febdb5c0149daa35926). I can add this fix there in router-great-again branch. Today evening.

**Comment by henri-hulski** [comment]
Great!

File: packages/router/README.md [hunk 5]
```diff
@@ -21,56 +21,184 @@ import {Controller} from 'cerebral'
...
+          map: {tab: props`tab`, focus: props`focus`},
+          signal: 'app.settingsRouted',
+        }
+      ]
+    })
+  }
+})
+```
+
+### routes definition (compute)
```
Comment (Guria) [inline packages/router/README.md]: Aren't we decided to delay this part?

**Comment by gaspard** [comment]
OK, I rebased and squashed some commits to cleanup all this. It should be ready to be merged in now.

**Pull request merged**